File: C++/056. Merge Intervals.cpp
#include <iostream>
#include <algorithm>
#include <vector>

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};

class Solution {
public:
    std::vector<Interval> merge(std::vector<Interval>& intervals) {
        std::vector<Interval> res;
        if (intervals.size() == 0)
            return res;
        sort(intervals.begin(), intervals.end(), compare);
        Interval interval(intervals[0]);
        for (int i = 1; i < intervals.size(); i++) {
            if (interval.end >= intervals[i].start) {
                if (interval.end < intervals[i].end)
                    interval.end = intervals[i].end;
            } else {
                res.emplace_back(interval);
                interval = intervals[i];
            }
        }
        res.emplace_back(interval);
        return res;
    }

    static bool compare(const Interval& a, const Interval& b) {
        return a.start < b.start;
    }
};

File: Go/056. Merge Intervals.go
package main

import "sort"

type Interval struct {
   Start int
   End   int
}

type IntervalSort []Interval

func (is IntervalSort)Len() int {
	return len(is)
}

func (is IntervalSort)Swap(i, j int) {
	is[i], is[j] = is[j], is[i]
}

func (is IntervalSort)Less(i, j int) bool {
	return is[i].Start < is[j].Start
}

func merge(intervals []Interval) []Interval {
	res := make([]Interval, 0)
	if len(intervals) == 0 {
		return res
	}
	sort.Sort(IntervalSort(intervals))
	interval := intervals[0]
	for i := 1; i < len(intervals); i++ {
		if intervals[i].Start <= interval.End {
			if intervals[i].End > interval.End {
				interval.End = intervals[i].End
			}
		} else {
			res = append(res, interval)
			interval = intervals[i]
		}
	}
	res = append(res, interval)
	return res
}